src/OeUI/OemrUI.php

```diff
@@ -82,39 +89,60 @@ public function __construct($arrOeUiSettings = array())
             $this->current_state = collectAndOrganizeExpandSetting($arrOeUiSettings['expandable_files']);
         }
+
+        $act = $this->arrAction;
+        $this->action = false;
+        // This is a holdover, the original code does not have a boolean for action, just a bunch of options, easier to keep for now
+        if ($act[0] != "") {
+            $this->action = true;
+        }
+
+        /**
+         * @var EventDispatcher
+         */
+        $this->ed = $GLOBALS['kernel']->getEventDispatcher();
+
+        /**
+         * @var TwigEnvironment
+         */
+        $twigContainer = new TwigContainer(null, $GLOBALS['kernel']);
+        $this->twig = $twigContainer->getTwig();
+
+        if ($this->expandable) {
+            $this->ed->addListener(PageHeadingRenderEvent::EVENT_PAGE_HEADING_RENDER, [$this, 'expandIconListener']);
+        }
+
+        if ($this->action) {
+            $this->ed->addListener(PageHeadingRenderEvent::EVENT_PAGE_HEADING_RENDER, [$this, 'actionIconListener']);
+        }
+
+        if ($GLOBALS['enable_help'] !== 0 && $this->display_help_icon) {
+            $this->ed->addListener(PageHeadingRenderEvent::EVENT_PAGE_HEADING_RENDER, [$this, 'helpIconListener']);
+        }
     }
 
     /**
-    * Creates the html string that will display the formatted heading with selected icons - expandable, action and help.
+    * Returns the page heading based on the options passed into the constructor.
     *
-    * @return array containing string $heading - the formatted html string of the actual heading and string $container
-    * - the value of the container class 'container' or 'container-fluid'
+    * @var bool $render If true, will echo the results from within this function
+    * @return string The Twig-rendered html content, suitable for simply echo'ing
     *
     */
-    public function pageHeading()
+    public function pageHeading(bool $render = false): null|string
     {
-        $heading = text($this->heading);
-        if (!empty($heading)) {
-            $arrexpandIcon = $this->expandIcon();// returns and array containing expandable icon string and container class string
-            $action_icon = $this->actionIcon();
-            $help_icon = $this->helpIcon();
-            $expandable_icon = $arrexpandIcon[0];
-            $heading = <<<HTML
-                <div class="d-flex">
-                    <div class="flex-grow-1">
-                        <h5 class="pb-0 mb-0 pt-1">$heading</h5>
-                    </div>
-                    <div class="pt-1">
-                        $expandable_icon
-                        $action_icon
-                        $help_icon
-                    </div>
-                </div>
-            HTML;
+        $pageHeadingEvent = $this->ed->dispatch(new PageHeadingRenderEvent(), PageHeadingRenderEvent::EVENT_PAGE_HEADING_RENDER);
+        $vars = [
+            "buttonList" => $pageHeadingEvent->getActions(),
+            "heading" => text($this->heading),
```

Comment:
much better to html escape as far downstream as possible for variables (ie. escape it in twig when it is outputted). This makes the code much easier to maintain security and avoid double escaping. Also then doesn't pigeon hole variables to certain escaping contexts (ie. if the twig template wishes to place the variable in a attr() context or a javascript context). There are several more variable below (such as title which is escaped prior to twig in 1 case and not the other also demonstrating why best to get it as downstream as possibe) which should also do this.